[Chandan8186] iP

Chandan8186 opened the pull request:
# ChatALot
## Your go-to task manager chatbot 🗒️
> There are two ways to write error-free programs: only the third one works. - Alan Perlis ([Source](https://www.brainyquote.com/quotes/alan_perlis_177353))

ChatALot is a command-line chatbot that
* is easy-to-**use**
* is easy to *set up*
* very ~~slow~~/efficient (for speedy typers)

## What can ChatALot do?
- [ ] Create/delete different types of tasks, with optional dates/timings
- [ ] Save the tasks that you have created, locally
- [x] Mark tasks as completed/not completed

## So what are you waiting for?
1. Download the latest version of the Jar file [here](https://github.com/Chandan8186/ip/releases/tag/Level-9)
2. Move it to your preferred directory
3. Open the command line and in the above directory, run `java -jar Duke.jar`
4. Chat away!

\
\
Psst...
Want a sneak peak of what the backend looks like?
```
.
.
.
private static final String HORIZONTAL_LINE
        = "_____________________________________________________\n";

private static final String INTRO_MESSAGE = HORIZONTAL_LINE
        + " ____  _   _   ____  _____  ____   _     ____  _____\n"
        + "/ (__`| |_| | / () \\|_   _|/ () \\ | |__ / () \\|_   _|\n"
        + "\\____)|_| |_|/__/\\__\\ |_| /__/\\__\\|____|\\____/  |_|\n\n"
        + "Hello! I'm ChatALot.\n"
        + "What can I do for you?\n"
        + HORIZONTAL_LINE;

public static void printIntro() {
        System.out.print(INTRO_MESSAGE);
}
.
.
.
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Chandan8186 committed:
Level-0: Rename, Greet, Exit

- Renamed Duke to ChatALot
- Added greet message (skeletal version)
- Added *.class to .gitignore

Chandan8186 committed:
Level-1: Echo

- ChatALot echoes user input
- minor design changes

Chandan8186 committed:
Level-2: Add, List

- functionality for storing items in a list

Chandan8186 committed:
Level-3: Mark as Done

- added task completion status

Chandan8186 committed:
Level-4: ToDo, Event, Deadline

- renamed ChatALot.java back to Duke.java
- added different task types

Chandan8186 committed:
A-TextUiTesting: Automated Text UI Testing

- important bug fixes
- added some tests for all commands

Chandan8186 committed:
Level-5: Handle Errors

- implemented error handling
- bug fixes
- updated runtest with more tests

Chandan8186 committed:
Level-6: Delete

- added delete functionality
- added test case for delete

Chandan8186 committed:
Add Save Functionality

- Fix some bugs and refactor code
- Add save functionality

Chandan8186 committed:
Merge branch 'branch-Level-7'

Chandan8186 committed:
Add date and time readability

- Fix some bugs
- Add date and time readability

Chandan8186 committed:
Update tests to accommodate date and time formats

- Change tests to test for invalid date and time formats

Chandan8186 committed:
Merge branch 'branch-Level-8'

Chandan8186 committed:
Add Ui and Storage classes

- add preliminary Ui and Storage classes integration

Chandan8186 committed:
Add Parser class

- Add parser class with parsing functionality
- Update saving functionality to save after each task list alteration
- Do some code refactoring and debugging

Chandan8186 committed:
Integrate all helper classes

- Integrate Parser, Storage and Ui with Duke
- Refactor some code

Chandan8186 committed:
Merge branch 'branch-A-MoreOOP'

Chandan8186 committed:
Organize into Packages

- Organize all non-trivial classes into duke, exceptions, tasks packages

Chandan8186 committed:
Merge branch 'branch-A-Packages'

Chandan8186 committed:
Merge commit 'ed6d4d25bd4e34461919ffe785a6f10f4c087552' into branch-A-Gradle

Chandan8186 committed:
Add gradle support

- Amend build.gradle accordingly

Chandan8186 committed:
Merge branch 'branch-A-Gradle'

Chandan8186 committed:
Add JUnit tests for tasks

- add some JUnit tests
- modify Parser to aid in testing

Chandan8186 committed:
Modify Storage class to take in file name

- Modify Storage class to take in any file name fed by user

Chandan8186 committed:
Merge branch 'branch-A-JUnit'

Chandan8186 committed:
Add javadocs

- Add javadocs for most classes and public methods

Chandan8186 committed:
Add more javadocs

- Add more javadocs

Chandan8186 committed:
Fix code to follow coding standards

- Fix coding standard issues

Chandan8186 committed:
Add find function to TaskList

- Add find function to TaskList

Chandan8186 committed:
Refactor a Ui function from level 9 branch

Chandan8186 committed:
Refactor Ui function and task list name from Javadoc branch

Chandan8186 committed:
Refactor Ui function from CodingStandards branch

Chandan8186 committed:
Implement finding task functionality

Chandan8186 committed:
Merge branch 'branch-A-JavaDoc'

Chandan8186 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Ui.java

Chandan8186 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/tasks/TaskList.java

Chandan8186 committed:
Fix merge conflict errors

- Fix error of erased javadocs from merge commit

Chandan8186 committed:
Configure checkstyle

- Add and configure checkstyle xml files

Chandan8186 committed:
Fix checkstyle and coding standard issues

- fix almost all checkstyle errors and warnings
- fix any coding standard issue
- refactored some method and attribute names

Chandan8186 committed:
Merge branch 'branch-A-CheckStyle'

Chandan8186 committed:
Implement rudimentary JavaFX gui

Chandan8186 committed:
Implement basic gui with full functionality

- add full functionality to basic JavaFX gui
- update build.gradle to allow for jar download

Chandan8186 committed:
Implement FXML

- implement FXML
- clean up / refactor / remove unnecessary code

Khaleelur-Rahman commented on src/main/java/Duke.java:
```diff
@@ -21,82 +21,159 @@ private static void intro() {
-        String output;
-        String desc;
+    private static String mark(String taskNumString) {
+        if (Duke.list.getSize() < 1) {
```
Good job in handling the empty list condition 👍

Khaleelur-Rahman commented on src/main/java/duke/Ui.java:
```diff
@@ -27,37 +28,37 @@ public static void outro() {
-                taskListSize +
-                " tasks in the list.");
+        System.out.println("Got it. I've added this task:\n"
+                + "  "
```
Good job in following the coding standards when wrapping statements :)

Khaleelur-Rahman replied:
Good job in ensuring that coding standards are followed when wrapping around statements :)

Khaleelur-Rahman commented on src/main/java/tasks/TaskList.java:
```diff
@@ -12,23 +12,23 @@
     protected static final DateTimeFormatter DATETIME_FORMAT
             = DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm");
-    protected ArrayList<Task> list;
+    protected ArrayList<Task> tasks;
```
Excellent work to pluralize your task list.

WeeeHung replied:
agreed!

Khaleelur-Rahman reviewed (commented):
Overall ,well done and keep up the good coding standards :)

Khaleelur-Rahman commented on src/main/java/tasks/Task.java:
```diff
@@ -5,30 +5,60 @@
     protected abstract String getTextFormattedString();
 
+    /**
+     * Constructor, initializes task description.
```
Good job in including javadocs for constructor methods too

Khaleelur-Rahman reviewed (commented):
Good to see javadocs for all the methods.

ZD292 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,152 @@
+package duke;
```
Looks good, everything is at same level of abstraction

ZD292 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,174 @@
+        }
+
+        String command = inputArr[0];
+        switch(command) {
```
I like how neat it is

ZD292 commented on src/main/java/tasks/DeadlineTask.java:
```diff
@@ -0,0 +1,45 @@
+ */
+public class DeadlineTask extends Task {
+
+    protected static final DateTimeFormatter INPUT_FORMAT =
```
Could this be private instead of protected?

Chandan8186 replied:
Seems like the task attributes could be private. Noted on the comments, thanks!

ZD292 reviewed (commented):
Overall really neat code that adheres to coding standard 🛩️

Chandan8186 committed:
Implement final GUI

- improve GUI
- refactor some code
- fix some bugs
- work on PR review comments

Chandan8186 committed:
Merge branch 'branch-Level-10'

WeeeHung commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,122 @@
+ */
+public class Storage {
+
+    private final String FILENAME;
```
i like that filename here is set as FINAL as we are only using a single file as local storage

WeeeHung commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,78 @@
+     * @return The formatted output for when a task is added
+     */
+    public static String getTaskAddingMessage(Task task, int taskListSize) {
+        return "Got it. I've added this task:\n"
```
perhaps string.format can make this cleaner!

Chandan8186 committed:
Add assertions for datetime objects

Some methods in some classes make important assumptions.

For example, when tasks with dates/times are created, the dates are assumed to not be null.

Adding assertions for the above makes sure that valid dates are printed to the user.

Chandan8186 committed:
Add assertions for filename and isMarked task attribute

Some methods in some classes make important assumptions.

For example, when tasks are created, the isMarked int is assumed
to be 0 (incomplete) or 1 (complete).

Adding assertions for the above makes sure that the complete/incomplete
cross is displayed to the user without issues.

Chandan8186 committed:
Improve readability for some methods

Some methods in Ui class were very long and difficult to read.

By using String.format(), the methods could be more concise
and readable.

This is in response to a PR review comment.

Chandan8186 committed:
Merge pull request #1 from Chandan8186/branch-A-Assertions

Implement A-Assertions

Chandan8186 committed:
Merge branch 'master' of https://github.com/Chandan8186/ip

Chandan8186 committed:
Merge pull request #2 from Chandan8186/branch-A-CodeQuality

Implement A-CodeQuality

Chandan8186 committed:
Merge branch 'master' of https://github.com/Chandan8186/ip

Chandan8186 committed:
Implement C-Statistics

Due to the many different types of tasks, users may find
it difficult to categorize them.

By giving task statistics, users can easily see the tasks due
the current day.

Chandan8186 committed:
Merge branch 'branch-BCD-Extension'

Chandan8186 committed:
Add gradle.yml to set up CI

Chandan8186 committed:
Merge branch 'branch-A-CI'

Chandan8186 committed:
Remove unused imports

Chandan8186 committed:
Fix minor checkstyle errors

- Fix checkstyle errors
- Suppress checkstyle error for final field name

Chandan8186 committed:
Update bye command

- Update bye command to close the program automatically

Chandan8186 committed:
Add user guide

- Update README to include user guide
- Add Ui screenshot

Chandan8186 committed:
Add some more javadocs

Chandan8186 committed:
Filter user input containing the vertical bar character

Chandan8186 committed:
Final changes before release

- Make some changes to user guide
- Release final jar file on github

Chandan8186 committed:
Amend compilation version

- Amend build.gradle to specify Java 11
- Re-release jar file
